math/src/main/kotlin/ktx/math/ImmutableVector.kt

```diff
@@ -0,0 +1,181 @@
+    /** Returns the result of adding the [other] vector to this vector */
+    operator fun plus(other: T): T
+
+    /** Returns the same vector with all members incremented by 1 */
+    operator fun inc(): T
+
+    /** Returns the same vector with all members decremented by 1 */
+    operator fun dec(): T
+
+    /** Returns this vector scaled by the given [scalar] */
+    operator fun times(scalar: Float): T
+
+    /** Returns his vector scaled by the given [vector] */
+    operator fun times(vector: T): T
+
+    /** Returns a vector of the same direction and a squared length of [length2] */
+    fun withLength2(length2: Float): T
+
+    /** Returns he dot product of this vector by the given [vector] */
+    infix fun dot(vector: T): Float
+
+    /**
+     * Returns the squared distance between this and the other vector
+     *
+     * This method is faster than [dst] because it avoids calculating a square root. It is useful for comparisons,
+     * but not for getting exact distance, as the return value is the square of the actual distance.
+     */
+    infix fun dst2(vector: T): Float
+
+    /** Linearly interpolates between this vector and the target vector by alpha */
+    fun withLerp(target: T, alpha: Float): T
+
+    /** Returns a vector of same length and a random direction */
+    fun withRandomDirection(): T
+
+    /** Returns a vector of same direction and the given [length] */
+    fun withLength(length: Float): T = withLength2(length * length)
+
+    /** Returns this vector if the [ImmutableVector.len] is <= [limit] or a vector with the same direction and length [limit] otherwise */
+    fun withLimit(limit: Float): T = withLimit2(limit * limit)
+
+    /** Returns this vector if the [ImmutableVector.len2] is <= [limit2] or a vector with the same direction and length [limit2] otherwise */
+    fun withLimit2(limit2: Float): T
+
+    /** Clamps this vector's length to given [min] and [max] values*/
+    fun withClamp(min: Float, max: Float): T = withClamp2(min * min, max * max)
+
+    /** Clamps this vector's squared length to given [min2] and [max2] values*/
+    fun withClamp2(min2: Float, max2: Float): T
+
+    /** Interpolates between this vector and the given [target] vector by [alpha] (within range [0,1]) using the given [interpolation] method. */
+    fun withInterpolation(target: T, alpha: Float, interpolation: Interpolation): T =
+            withLerp(target, interpolation.apply(alpha))
+
+    /** Returns true if this vector is in line with the other vector (either in the same or the opposite direction) */
+    fun isOnLine(other: T, epsilon: Float = MathUtils.FLOAT_ROUNDING_ERROR): Boolean
+
+    /** Compares this vector with the other vector, using the supplied epsilon for fuzzy equality testing */
+    fun epsilonEquals(other: T, epsilon: Float): Boolean
+
+    override fun compareTo(other: T): Int = len2.compareTo(other.len2)
+
+    @Deprecated(MUTABLE_METHOD_DEPRECATION_MESSAGE, ReplaceWith("withLerp(target, alpha)"))
```

Comment:
`level = DeprecationLevel.ERROR`

Better play it safe and be a little annoying to the users than cause potential bugs.

Reply:
Same goes for all mutating methods.

Reply:
Yes, indeed.